profiling/src/profile/mod.rs

```diff
@@ -221,16 +342,67 @@ struct StackTrace {
     pub locations: Vec<LocationId>,
 }
 
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash, PartialOrd, Ord)]
+enum LabelValue {
+    Str(StringId),
+    Num {
+        num: i64,
+        num_unit: Option<StringId>,
+    },
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash, PartialOrd, Ord)]
+struct Label {
+    key: StringId,
+    value: LabelValue,
+}
+
+impl Label {
+    pub fn str(key: StringId, v: StringId) -> Self {
+        Self {
+            key,
+            value: LabelValue::Str(v),
+        }
+    }
+
+    pub fn num(key: StringId, num: i64, num_unit: Option<StringId>) -> Self {
+        Self {
+            key,
+            value: LabelValue::Num { num, num_unit },
+        }
+    }
+}
+
+impl From<Label> for pprof::Label {
+    fn from(l: Label) -> pprof::Label {
+        let key = l.key.into();
+        match l.value {
+            LabelValue::Str(str) => Self {
+                key,
+                str: str.into(),
+                num: 0,
+                num_unit: 0,
+            },
+            LabelValue::Num { num, num_unit } => Self {
+                key,
+                str: 0,
+                num,
+                num_unit: num_unit.map(i64::from).unwrap_or(0),
+            },
+        }
+    }
+}
+
 #[derive(Eq, PartialEq, Hash)]
 struct Sample {
     pub stacktrace: StackTraceId,
 
     /// label includes additional context for this sample. It can include
     /// things like a thread id, allocation size, etc
-    pub labels: Vec<Label>,
+    pub labels: LabelSetId,
 
     /// Offset into `labels` for the label with key == "local root span id".
```

Comment:
(The old version of the comment is now outdated)

```suggestion
    /// This label is stored separately, to make it easy to add the endpoint label during serialization.
```